### PR TITLE
feat(model): setup data preprocessing pipeline

### LINKED ISSUE #75: [Super AI] Setup Data Preprocessing Pipeline
Description: We need a data preprocessing pipeline as a script that **convert Huggingface text dataset to Hugginface tokenized id datasets** https://huggingface.co/docs/datasets/index

The scripts should have at least but not limited to these arguments:
- str: sentencepiece tokenizer (*.spm) or huggingface (tokenizer.json) tokenizer model path (automatically select based on pipeline)
- str: dataset input path
- str: dataset save path
- bool: output dataset should have mask_id or not

The Scripts should work with
- Multi CPUs
- 1TB++ of Data without crashing memory (OOM)

Examples in smaller scales (XX Gb data size):
- https://github.com/OpenThaiGPT/Pretraining-ELECTRA/blob/main/src/electra_pretraining/cli_tools/datasets_preparation.py

Deliverables:
- Scripts
- Pull Request

### PR DESCRIPTION
**Pull request by pitikorn32**
## Why this PR
We need a data preprocessing pipeline as a script that convert Huggingface text dataset to Hugginface tokenized id datasets

## Changes
- add `src\model\openthaigpt_pretraining_model\tokenizers\data_preprocessing.py`

## Related Issues
Close #75

### CODE REVIEW

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        self.tokenizer.load(tokenizer_model_path)
+
+    @staticmethod
+    def preprocess_text(text: str) -> str:
+        # Custom
+        return text
+
+    # define a function to tokenize the text
+    def tokenize_text(self, text: str) -> str:
+        text = SPMTokenizer.preprocess_text(text)
```
[the remark this answers is not in the document]
Comment (pitikorn32) [reply src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py]: Fixed

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            text = f.read()
+        yield {"text": text}
+
+
+def tokenize_dataset(
+    tokenizer_model_path: str,
+    output_path: str,
+    num_docs: Optional[Union[str, int]] = 5000,
+    num_proc: Optional[int] = os.cpu_count(),
+    batch_size: int = 1000,
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py:44]: We can increase this number to 100_000

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                split=f"train[:{num_docs}]",
+            )
+
+    else:
+        # create a dataset from the generator
+        text_dataset = Dataset.from_generator(
+            text_generator, gen_kwargs={"data_path": load_dataset_local_path}
+        )
+
+    tokenizer = SPMTokenizer(tokenizer_model_path)
```
[the remark this answers is not in the document]
Comment (pitikorn32) [reply src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py]: yes, I added `FastTokenizerWrapper` for Fast tokenizer.

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py (hunk 2 above)
Comment (pitikorn32) [reply src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py]: Okay, thank

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        # Custom
+        return text
+
+    # define a function to tokenize the text
+    def tokenize_text(self, text: str) -> str:
+        text = self.preprocess_text(text)
+        return self.tokenizer.encode(text, out_type=int)
+
+
+def text_generator(data_path: str) -> Generator[Dict[str, str], None, None]:
```
[the remark this answers is not in the document]
Comment (pitikorn32) [reply src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py]: I'm not sure. I think generator can handle OOM and in the case if individual text file is too large I also add `chunk_generator` to chunk in each text file.

File: src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        text = self.preprocess_text(text)
+        return self.tokenizer.encode(text, out_type=int)
+
+
+class FastTokenizerWrapper(BaseTokenizer):
+    def __init__(self, tokenizer_model_path: str) -> None:
+        super().__init__(tokenizer_model_path)
+
+        # load the tokenizer model
+        self.tokenizer = AutoTokenizer.from_pretrained(self.tokenizer_model_path)
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py:45]:
```suggestion
        self.tokenizer = PreTrainedTokenizerFast.from_pretrained(self.tokenizer_model_path)
```

Comment (pitikorn32) [reply src/model/openthaigpt_pretraining_model/tokenizers/data_preprocessing.py]: Revised.

**Review by new5558 — APPROVED** [review]

**Pull request merged**